hi-ml/src/health_ml/utils/generic_parsing.py

```diff
@@ -0,0 +1,451 @@
+
+# Need this as otherwise a description of all the params in a class is added to the class docstring
+# which makes generated documentation with sphinx messy.
+param.parameterized.docstring_signature = False
+param.parameterized.docstring_describe_params = False
+
+
+@dataclass
+class ParserResult:
+    """
+    Stores the results of running an argument parser, broken down into a argument-to-value dictionary,
+    arguments that the parser does not recognize.
+    """
+    args: Dict[str, Any]
+    unknown: List[str]
+    overrides: Dict[str, Any]
+
+
+class ListOrDictParam(param.Parameter):
+    """
+    Wrapper class to allow either a List or Dict inside of a Parameterized object.
+    """
+
+    def _validate(self, val: Any) -> None:
+        if not (self.allow_None and val is None):
+            if not (isinstance(val, List) or isinstance(val, Dict)):
+                raise ValueError(f"{val} must be an instance of List or Dict, found {type(val)}")
+        super()._validate(val)
+
+
+class StringOrStringList(param.Parameter):
+    """
+    Wrapper class to allow either a string or a list of strings. Internally represented always as a list.
+    """
+
+    def _validate(self, val: Any) -> None:
+        if isinstance(val, str):
+            return
+        if isinstance(val, List):
+            if all([isinstance(v, str) for v in val]):
+                return
+        raise ValueError(f"{val} must be a string or a list of strings")
+
+    def set_hook(self, obj: Any, val: Any) -> Any:
+        """
+        Modifies the value before calling the setter. Here, we are converting all strings to lists of strings.
+        """
+        if isinstance(val, str):
+            return [val]
+        return val
+
+
+class PathOrPathList(param.Parameter):
+    """
+    Wrapper class to allow either a Path or a list of Paths. Internally represented always as a list.
+    """
+
+    def _validate(self, val: Any) -> None:
+        if isinstance(val, Path):
+            return
+        if isinstance(val, List):
+            if all([isinstance(v, Path) for v in val]):
+                return
+        raise ValueError(f"{val} must be a Path object or a list of paths")
+
+    def set_hook(self, obj: Any, val: Any) -> Any:
+        """
+        Modifies the value before calling the setter. Here, we are converting simple path to lists of path.
+        """
+        if isinstance(val, Path):
+            return [val]
+        return val
+
+
+class IntTuple(param.NumericTuple):
+    """
+    Parameter class that must always have integer values
+    """
+
+    def _validate(self, val: Any) -> None:
+        super()._validate(val)
+        if val is not None:
+            for i, n in enumerate(val):
+                if not isinstance(n, int):
+                    raise ValueError("{}: tuple element at index {} with value {} in {} is not an integer"
+                                     .format(self.name, i, n, val))
+
+
+class GenericConfig(param.Parameterized):
```

Comment:
I think GenericConfig is superseded by the argument parsers that you wrote the commandline tools